### PR TITLE
Mobile/enhancement/overlapped annotations 545

### LINKED ISSUE #545: Mobile: Implementing Overlapped Annotation Support
### Issue Description

I have already implemented [annotatable text](https://github.com/bounswe/bounswe2022group2/issues/505) and [highlighted/annotated text](https://github.com/bounswe/bounswe2022group2/issues/517) widgets in the previous weeks. However, according to our discussiona with the TA and Suzan hoca, we need to support overlapped annotated texts also. Users shall be able annotate the texts whose some parts (partially) have already annotated by someone else. Therefore, I will add overlapping support to the annotatable and annotated text widgets with this issue.

### Step Details

Steps that will be performed:
- [x] Remove the logic that disables the annotatability of the already annotated parts
- [x] Add support for choosing and being able to annotate any parts of the text
- [x] Design and implement an algorithm to divide the annotated texts into separate parts to detect the texts which are annotated more than once
- [x] Integrate new color and its annotation fields into the annotation model
- [x] Implement a color helper mixin, implement helper functions for creating a random color and creating a random light color
- [x] Implement a color mixture helper method for overlapped annotations

### Final Actions

I will test the implementation and widgets created on my real phone. If everything works like a charm, I will open a pull request. If the pull request is approved by at least one teammate, I will merge it and close the issue.

### Deadline of the Issue

20.11.2022 - Sunday [23.59]

### Reviewer

Egemen Atik

### Deadline for the Review

21.11.2022 - Monday [15.00]

### PR DESCRIPTION
**Pull request by bahricanyesil**
## Short Description

This PR adds overlapping support to the current implementation of the annotatable text. Now, users will be able to annotate any part of the text without dealing with conflicts. Also, annotations now have different colors to differentiate them.

------

## What Does This PR Include?

- [x] Removes the logic that disables the annotatability of the already annotated parts
- [x] Adds support for choosing and being able to annotate any parts of the text
- [x] Implements the algorithm to divide the annotated texts into separate parts to detect the texts which are annotated more than once
- [x] Integrates new color and its annotation fields into the annotation model
- [x] Implements a color helper mixin, implements helper functions for creating a random color and creating a random light color
- [x] Implements a color mixture helper method for overlapped annotations

------

Closes #545.

### CODE REVIEW

**Comment by bahricanyesil** [comment]
Here's a video recording showing the support for overlapping annotations:

https://user-images.githubusercontent.com/60237280/202254626-704c80ad-b1de-4511-8e88-183ad60580f1.mov

File: learnify/mobile/lib/features/auth/login/view-model/login_view_model.dart [hunk 1]
```diff
@@ -71,6 +71,16 @@ class LoginViewModel extends BaseViewModel {
   }
 
   Future<String?> _loginRequest() async {
+    // TODO:
+    const User user = User(id: '2313');
+    await localManager.setModel(user, StorageKeys.user);
+    await localManager.setString(StorageKeys.accessToken, 'toke');
```
Comment (ezgy) [inline learnify/mobile/lib/features/auth/login/view-model/login_view_model.dart:77]: I think you have a typo here with 'token'

File: learnify/mobile/lib/features/auth/login/view-model/login_view_model.dart [hunk 2]
```diff
@@ -71,6 +71,16 @@ class LoginViewModel extends BaseViewModel {
...
+    await localManager.setInt(StorageKeys.accessTokenExpires,
+        DateTime.now().add(const Duration(days: 10)).millisecondsSinceEpoch);
+    await navigationManager.navigateToPageClear(
```
Comment (ezgy) [inline learnify/mobile/lib/features/auth/login/view-model/login_view_model.dart:80]: I understand the addition of token expiration, but isn't the rest repeating line 104+ with no changes and no deletion?

File: learnify/mobile/lib/features/auth/login/view-model/login_view_model.dart [hunk 3]
```diff
@@ -71,6 +71,16 @@ class LoginViewModel extends BaseViewModel {
   }
 
   Future<String?> _loginRequest() async {
+    // TODO:
+    const User user = User(id: '2313');
```
Comment (ezgy) [inline learnify/mobile/lib/features/auth/login/view-model/login_view_model.dart:75]: this shouldn't be here?

**Review by ezgy — CHANGES_REQUESTED** [review]
The overlapped annotation seems amazing, great work! I have a few questions about the changes in login_view_model though.

**Comment by bahricanyesil** [comment]
> The overlapped annotation seems amazing, great work! I have a few questions about the changes in login_view_model though.

Thank you!

**Review by ezgy — APPROVED** [review]
Seems good to merge.

**Pull request merged**